design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,302 @@
+can use `INSERT INTO ... VALUES` to quickly construct a toy environment for testing. Generating this 
+data KSQL, however, is cumbersome: 
+
+* you require a separate tool (e.g. `kafkacat`, `ksql-datagen`)
+* you must know the deployment topology (e.g. URIs for ZK/Broker)
+* there is no synergy with KSQL (e.g. tools are not *schema aware* and can produce arbitrary data)
+
+This KLIP attempts to integrate producing test data into the KSQL REPL.
+
+## Scope
+
+This KLIP is intended to improve the *development and testing* workflow as opposed to providing
+*production-ready* functionality. The following tasks will be addressed:
+
+* KSQL will be able to create and produce data to arbitrary topics
+* KSQL will be able to validate data when producing to a topic that is registered with a schema
+in KSQL (e.g. as a result of `CREATE STREAM`)
+
+## Value
+
+This proposal will improve the first 15-minute experience that developers have with KSQL and will
+help seasoned KSQL developers test new functionality. Furthermore, this functionality can be
+leveraged to create a powerful testing harness for KSQL applications that can run in a sandbox
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO
+
+```sql
+INSERT INTO <stream_name|table_name> [(column1, column2, ...)]
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT INTO` statement can be used to insert new records into a kafka topic (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. `ROWKEY` and `ROWTIME` are always present as the first
+two columns, and are automatically populated with semantics described in the section on `ROWKEY` and
+`ROWTIME` semantics below.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (see _Complex Datatypes_
+section below).
+
+Future extensions: 
+* Support some`AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+
+#### DELETE FROM
+
+```sql 
+DELETE FROM table_name WHERE ROWKEY = <rowkey_value>;
+```
+
+The `DELETE` statement can be used to insert tombstones into kafka topics that back tables. If the
+source is not back a table, the statement will fail. For this KLIP, we will only support conditions
+of the form `ROWKEY = value`. While this may initially be confusing, maintaining consistent
+documentation and examples, along with adequate error messages, will alleviate the concern.
+
+#### CREATE STREAM/TABLE
+
+```sql
+CREATE STREAM stream_name ( { column_name data_type } [, ...] )
```

Comment:
From what I can tell, our story around avro schema validation is... iffy... Maybe I'm wrong but the code is a little hard to follow. It doesn't look like we require the corresponding Schema to exist in SR if you supply the schema anyway (if you don't supply a schema, and it doesn't exist, then it fails):
```
ksql> CREATE STREAM pageviews_avro (pageid varchar, viewtime bigint, userid varchar) WITH (kafka_topic='pageviews', value_format='avro');

 Message
----------------
 Stream created
----------------
ksql> CREATE STREAM pageviews_avro2  WITH (kafka_topic='pageviews', value_format='avro');
Avro schema for message values on topic pageviews does not exist in the Schema Registry.Subject: pageviews-value
Possible causes include:
- The topic itself does not exist	-> Use SHOW TOPICS; to check
- Messages on the topic are not Avro serialized	-> Use PRINT 'pageviews' FROM BEGINNING; to verify
- Messages on the topic have not been serialized using the Confluent Schema Registry Avro serializer	-> See https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html
- The schema is registered on a different instance of the Schema Registry	-> Use the REST API to list available subjects	https://docs.confluent.io/current/schema-registry/docs/api.html#get--subjects
```

Given the current rather inconsistent behavior, I'll go down the path of creating a schema in SR. Since we won't allow creating a topic that already exists, we shouldn't run into the issue of having a CT that doesn't match the one in SR (at least any worse than we have it now).